test: raft: replace the use of read_barrier work-around

emdroid opened the pull request:
Replaced the old `read_barrier` helper from "test/pylib/util.py" by the new helper from "test/pylib/rest_client.py" that is calling the newly introduced direct REST API.

Replaced in all relevant tests and decommissioned the old helper.

Introduced a new helper `get_host_api_address` to retrieve the host API address - which in come cases can be different from the host address (e.g. if the RPC address is changed).

Fixes: scylladb/scylladb#19662

emdroid commented:
Note to reviewers:
* the PR currently includes the #19715 to avoid test failures - please only review the second commit relevant to this change
* the PR will be rebased later after the #19715 is merged to only include the relevant commit(s)

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10395/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10395/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/2600/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/10619/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10395/)
#### Failed Tests (10/31399):
- [alternator.test_scan.dev.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/10395/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___alternator_test_scan_dev_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+alternator.test_scan.dev.3)
- [test_change_rpc_address.debug.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/10395/testReport/junit/%28root%29/test_change_rpc_address/Tests___Unit_Tests___test_change_rpc_address_debug_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_change_rpc_address.debug.1)
- [test_change_rpc_address.debug.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/10395/testReport/junit/%28root%29/test_change_rpc_address/Tests___Unit_Tests___test_change_rpc_address_debug_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_change_rpc_address.debug.2)
- [test_change_rpc_address.debug.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/10395/testReport/junit/%28root%29/test_change_rpc_address/Tests___Unit_Tests___test_change_rpc_address_debug_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_change_rpc_address.debug.3)
- [test_change_rpc_address.dev.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/10395/testReport/junit/%28root%29/test_change_rpc_address/Tests___Unit_Tests___test_change_rpc_address_dev_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_change_rpc_address.dev.1)
- [test_change_rpc_address.dev.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/10395/testReport/junit/%28root%29/test_change_rpc_address/Tests___Unit_Tests___test_change_rpc_address_dev_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_change_rpc_address.dev.2)
- [test_change_rpc_address.dev.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/10395/testReport/junit/%28root%29/test_change_rpc_address/Tests___Unit_Tests___test_change_rpc_address_dev_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_change_rpc_address.dev.3)
- [test_change_rpc_address.release.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/10395/testReport/junit/%28root%29/test_change_rpc_address/Tests___Unit_Tests___test_change_rpc_address_release_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_change_rpc_address.release.1)
- [test_change_rpc_address.release.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/10395/testReport/junit/%28root%29/test_change_rpc_address/Tests___Unit_Tests___test_change_rpc_address_release_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_change_rpc_address.release.2)
- [test_change_rpc_address.release.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/10395/testReport/junit/%28root%29/test_change_rpc_address/Tests___Unit_Tests___test_change_rpc_address_release_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_change_rpc_address.release.3)
#### Build Details:
- Duration: 5 hr 46 min
- Builder: i-0ab2d5371effb4f51 (m5ad.8xlarge)

emdroid commented:
New version:
* rebased
* fixed the failing RPC tests

In some cases the host "base" address can be different than the API (listening) address, in particular when the RPC address is being changed (which is the case of the failing "change_rpc_address" tests). Added the `get_host_api_address` helper to determine the correct host API address to use.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10404/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10404/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/10404/)
#### Build Details:
- Duration: 2 hr 22 min
- Builder: spider8.cloudius-systems.com
